### PR TITLE
Generalize variable

### PR DESCRIPTION
**Pull request by Codoscope**
PR corresponding to the issue [https://github.com/pasqal-io/Pulser/issues/325](https://github.com/pasqal-io/Pulser/issues/325)

I still have to fix a problem when trying to rename the classes, falling on a `TypeError: 'ABCMeta' object is not subscriptable` error.

### CODE REVIEW

**Comment by HGSilveri** [comment]
> I still have to fix a problem when trying to rename the classes, falling on a `TypeError: 'ABCMeta' object is not subscriptable` error.

Ok, I'll mark it as a Draft for now then

**Comment by HGSilveri** [comment]
Is this ready for review or are you still going to rename the `Variable` and `_VariableItem` classes?

File: pulser/sequence.py [hunk 1]
```diff
@@ -538,12 +548,31 @@ def declare_channel(
...
         dtype: Union[type[int], type[float], type[str]] = float,
     ) -> Variable:
+        pass
+
+    def declare_variable(
+        self,
+        name: str,
+        size: Optional[int] = None,
+        dtype: Union[type[int], type[float], type[str]] = float,
+    ) -> Union[Variable, _VariableItem]:
```
Comment (HGSilveri) [inline pulser/sequence.py:575]: I like this solution, I think it's better than having two separate methods.

File: pulser/parametrized/paramobj.py [hunk 2]
```diff
@@ -56,6 +56,48 @@ def __neg__(self) -> ParamObj:
...
+        pass
+
+    def __rpow__(self, other: Union[int, float]) -> ParamObj:
+        pass
+
+    def __mod__(self, other: Union[int, float]) -> ParamObj:
+        pass
+
+    def __rmod__(self, other: Union[int, float]) -> ParamObj:
+        pass
```
[the remark this answers is not in the document]
Comment (Codoscope) [reply pulser/parametrized/paramobj.py:99]: The benefit of doing this is that the implementation is still dynamically factorized. Only the redundancy of declarations necessary to mypy is added. But I can also include the implementation if you think it's better.

[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser/parametrized/paramobj.py:99]: In your opinion, what's the benefit of having it dynamically factorized?

[the remark this answers is not in the document]
Comment (Codoscope) [reply pulser/parametrized/paramobj.py:99]: It avoids the burden of repeating `ParamObj(operator.__add__, self, other)`, `ParamObj(operator.__radd__, other, self)` and so on in all functions. The repetition is pretty small however, so it's true that the dynamic code might not be worth it anymore.

[the remark this answers is not in the document]
Comment (HGSilveri) [reply pulser/parametrized/paramobj.py:99]: To me, it won't add any lines and it will make the code more explicit, so I say go for it.

File: pulser/tests/test_paramseq.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
 def test_var_declarations():
     sb = Sequence(reg, device)
     assert sb.declared_variables == {}
-    var = sb.declare_variable("var")
+    var = sb.declare_variable("var", size=1)
```
Comment (HGSilveri) [inline pulser/tests/test_paramseq.py:32]: Maybe you could add a test where the variable is declared with `size=None` too

File: pulser/tests/test_paramseq.py [hunk 4]
```diff
@@ -175,7 +175,8 @@ def test_str():
     sb.add(pls, "ch1")
     s = (
         f"Prelude\n-------\n{str(seq)}Stored calls\n------------\n\n"
-        + "1. add(Pulse.ConstantPulse(mul(var, 100), var, -1, var), ch1)"
+        + "1. add(Pulse.ConstantPulse(mul(var[0], 100), var[0],"
```
Comment (HGSilveri) [inline pulser/tests/test_paramseq.py:178]: I guess that it would be nice that a variable declared with `size=None` would keep being described without the `[0]`, but that's a small detail.

**Review by HGSilveri — APPROVED** [review]
LGTM, I only have a couple of minor remarks

File: pulser/tests/test_paramseq.py (hunk 3 above)
Comment (Codoscope) [reply pulser/tests/test_paramseq.py:32]: This corresponds to the line `var = sb.declare_variable("var")` since `size` is defaulted to `None`. Do you think I should add a test where it is explicitely `None` too?

File: pulser/tests/test_paramseq.py (hunk 4 above)
Comment (Codoscope) [reply pulser/tests/test_paramseq.py:178]: Do you think we could leave it for another PR and keep the Issue open? I even wonder if the `_VariableItem` class shouldn't be removed in favor of an instanciation of `ParamObj` doing the indexation. I have some draft about it but it's having UT issues.

File: pulser/tests/test_paramseq.py (hunk 3 above)
Comment (HGSilveri) [reply pulser/tests/test_paramseq.py:32]: Yes, but you changed it to have `size=1` so now I don't know if there is one where size was not specified (ie defaulted to `None`). No need for a test where `None` is explicitly stated

File: pulser/tests/test_paramseq.py (hunk 4 above)
Comment (HGSilveri) [reply pulser/tests/test_paramseq.py:178]: Sure, that's okay with me

File: pulser/tests/test_paramseq.py (hunk 3 above)
Comment (Codoscope) [reply pulser/tests/test_paramseq.py:32]: I meant that the line `var = sb.declare_variable("var")` is already present in other parts of the test file. For example, on line 134 in `test_build()`

Comment (HGSilveri) [reply pulser/tests/test_paramseq.py:32]: Sure, of course. But I think it would be worth it to have the checks that are done after this declaration for this case too. For example:
```python
var_ = seq.declare_variable("var_")
assert sb.declared_variables["var_"] == var_
assert isinstance(var_, _VariableItem)
```
You see what I mean?

Comment (Codoscope) [reply pulser/tests/test_paramseq.py:33]: Yes, thanks for your remark, I have applied it, although a downside of this approach in my sense is that it seems to be testing the implementation instead of the behavior.

**Review by HGSilveri — APPROVED** [review]
Thanks a lot @Codoscope !
